pgptool-gui/src/main/java/org/pgptool/gui/tools/fileswatcher/MultipleFilesWatcher.java

```diff
@@ -36,230 +36,87 @@
-				if (baseFolder == null) {
-					log.debug("No associated watchers found for " + baseFolderStr);
-					return;
+				@Override
+				public void onFileCreate(File file) {
+					dirWatcherHandler.handleFileChanged(ENTRY_CREATE, file.getAbsolutePath());
 				}
 
-				baseFolder.interestedFiles.remove(relativeFilename);
-				log.debug("File is no longer watched in folder " + baseFolderStr + " file " + relativeFilename);
-				if (baseFolder.interestedFiles.size() > 0) {
-					return;
+				@Override
+				public void onFileDelete(File file) {
+					dirWatcherHandler.handleFileChanged(ENTRY_DELETE, file.getAbsolutePath());
 				}
+			});
 
-				// NOTE: Decided to turn this off, because file might re-appear in case it's app
-				// re-created it. See #91, #75
-				// keys.remove(baseFolder.key);
-				// baseFolders.remove(baseFolder.folder);
-				// baseFolder.key.cancel();
-				// log.debug("Folder watch key is canceled " + baseFolderStr);
-			}
-		} catch (Throwable t) {
-			log.error("Failed to watch file " + filePathName, t);
-			// not goinf to ruin app workflow
+			monitor.addObserver(observer);
+			observers.put(baseFolderStr, observer);
 		}
 	}
 
-	private Thread buildThreadAndStart() {
-		Thread ret = new Thread("FilesWatcher-" + watcherName) {
-			@Override
-			public void run() {
-				log.debug("FileWatcher thread started " + watcherName);
-				boolean continueWatching = true;
-				while (continueWatching) {
-					WatchKey key;
-					try {
-						idleIfNoKeysRegistered();
-						key = watcher.take();
-						// NOTE: Since we're watching only one folder we assume
-						// that there will be only one key for our folder
-					} catch (ClosedWatchServiceException cwe) {
-						log.info("ClosedWatchServiceException fired, stopping thread.");
-						return;
-					} catch (InterruptedException x) {
-						log.debug("FileWatcher thread stopped by InterruptedException");
-						return;
-					} catch (Throwable t) {
-						log.error("Unexpected exception while checking for updates on watched file", t);
-						return;
-					}
-
-					BaseFolder baseFolder = null;
-					synchronized (watcherName) {
-						baseFolder = keys.get(key);
-					}
-					if (baseFolder == null) {
-						key.cancel();
-						continue;
-					}
-
-					for (WatchEvent<?> event : key.pollEvents()) {
-						// Context for directory entry event is the file name of
-						// entry
-						WatchEvent<Path> ev = cast(event);
-						Path name = ev.context();
-						Path child = baseFolder.path.resolve(name);
-						String relativeFilename = FilenameUtils.getName(child.toString());
-						if (!baseFolder.interestedFiles.contains(relativeFilename)
-								&& !event.kind().equals(ENTRY_CREATE)) {
-							continue;
-						}
-
-						// print out event
-						log.debug("Watcher event: " + event.kind().name() + ", file " + child);
-						dirWatcherHandler.handleFileChanged(event.kind(), child.toString());
-					}
+	public void stopWatchingFile(String filePathName) {
```

Comment:
Impl of this class and especially of this method doesn't match expectations.

Steps To Reproduce

1. Decrypt fileA.pgp to folderA/fileA
2. Decrypt fileB.pgp to folderA/fileB
3. Use any other application to delete fileA
4. fileA disappears from the "Previously decrypted" files list in PGP Tool
5. Now use any other application to delete fileB

**Actual behavior**: fileB remains in the "Previously decrypted" files list in PGP Tool

**Expected behavior**: fileB must also disappear from the "Previously decrypted" files list in PGP Tool

**How to fix**: you need to track files per folder and stop per-folder-monitors only when no files left to be monitored in this folder.